Fix no warning when leaving edited native questions

WiNloSt opened the pull request:
Closes https://github.com/metabase/metabase/issues/31974
Epic: https://github.com/metabase/metabase/issues/29421

### Description

The previous approach uses `isNew` in this condition `!isNew && (isExistingModelDirty || isExistingSqlQueryDirty)` at glance it reads like we're making sure we only return `true` when we're editing existing questions hence `!isNew`, however `isNew` was introduced in https://github.com/metabase/metabase/pull/2795 which returns true when `state.qb.card.id` is not there which is the case when editing existing native questions too.

The new approach uses logic found in `frontend/src/metabase/containers/SaveQuestionModal.tsx`–which is used to determine when to display options to replace the original question or save it as a new question.
![image](https://github.com/metabase/metabase/assets/1937582/6cd9aa65-9ff7-4dc2-aecb-378b4017ad1b)

And logic in `frontend/src/metabase/selectors/app.ts`–which is used to determine when to show a question lineage
![image](https://github.com/metabase/metabase/assets/1937582/87787c25-8dc6-4a51-a648-e217ddc3f179)

Both logic represents the same thing. To check if a question–not a model–has been modified.

### How to verify

1. Go to an existing SQL question
2. Edit the query (ex: click `Open Editor` and add/delete a character in the SQL query)
3. Try to refresh/close tab/go to a different URL, you should see a message similar to that of the video below.

### Demo
https://www.loom.com/share/45a7c05952f74721ba2629c52a9fdaf9

### Checklist

- [x] Tests have been added/updated to cover changes in this PR.

WiNloSt committed:
Add warning message when leaving an edited, existing SQL question

This PR adds this functionality that's supposed be a part of
an epic https://github.com/metabase/metabase/issues/29421

WiNloSt committed:
Use similar structure on other part that rely on the same logic

WiNloSt commented on frontend/src/metabase/query_builder/selectors.js:
```diff
@@ -180,8 +181,6 @@ export const getPKRowIndexMap = createSelector(
   },
 );
 
-export const getIsNew = state => state.qb.card && !state.qb.card.id;
```
This selector was added in https://github.com/metabase/metabase/pull/2795 and was used in `frontend/src/metabase/query_builder/QueryHeader.jsx` which doesn't seem to exist anymore. I also try to find other part of the code that would relate to that file which still uses `isNew`, but couldn't find anything.

WiNloSt commented on frontend/src/metabase/query_builder/utils.ts:
Convert the whole file to TypeScript so I could use type predicate (`originalQuestion is Question`) in `isQuestionEdited` to avoid TypeScript from complaining that `originalQuestion` could be null when used in
https://github.com/metabase/metabase/blob/5f94e08ccddd674d299ef448f9810addfdb99ce1/frontend/src/metabase/containers/SaveQuestionModal.tsx#L143-L145

ranquild commented on frontend/src/metabase/query_builder/containers/QueryBuilder.jsx:
```diff
@@ -300,18 +303,16 @@ function QueryBuilder(props) {
+  const shouldShowUnsavedChangesWarningForModels =
+    isEditingModel && (isDirty || isMetadataDirty);
+  const shouldShowUnsavedChangesWarningForSqlQuery =
+    isNativeEditorOpen && isQuestionEdited;
```
Why do we check that the native editor is opened (expanded) here? It doesn't change the fact my changes are lost. Should we check that the question is native (`question.isNative()`) instead?

WiNloSt replied:
That's a good point. I used existing logic which also uses `isNativeEditorOpen`, but I can see if we open the editor edit the query then close the editor, there would be no browser prompt when we refresh the page.

ranquild commented on frontend/src/metabase/query_builder/containers/QueryBuilder.jsx:
```diff
@@ -300,18 +303,16 @@ function QueryBuilder(props) {
-    () => isModelQueryDirty || isMetadataDirty,
-    [isMetadataDirty, isModelQueryDirty],
-  );
+  const shouldShowUnsavedChangesWarningForModels =
```
Wdyt about extracting this logic in a separate hook (in this file) and use `useSelector` instead of adding more props to the root component?

WiNloSt replied:
sounds good to me.

ranquild reviewed (commented):
Comments inline

WiNloSt commented on frontend/src/metabase/query_builder/utils.ts:
```diff
@@ -39,3 +58,15 @@ export function getURLForCardState({ card }, dirty, query = {}, objectId) {
+  return (
+    question != null &&
+    !question.isSaved() &&
+    originalQuestion != null &&
```
This is the logic that's extracted from `frontend/src/metabase/selectors/app.ts` and `frontend/src/metabase/containers/SaveQuestionModal.tsx`. They're technically the same.

I think we need to rely on `originalQuestion.isDataset()` because when you edit a model it won't prompt you to replace the model, unlike editing a question which will ask you to either modify the saved question or save it as a new question instead. i.e. if we only check `question.isSave()` we will ask users whether they want to save a modified model but we don't want that this is what we should show
![image](https://github.com/metabase/metabase/assets/1937582/c24d7463-ac8b-4b85-a377-b4b9c037f0d2)

The same thing applies to the question lineage we won't show question lineage for models
![image](https://github.com/metabase/metabase/assets/1937582/880b0f3a-08fb-4cb5-96c7-dc493bff6a93)

This logic is extracted here for the purpose of those 2 use cases, but it is used in this fix as well.

Now I've explained the reason that this logic exists. And the reason it's used for this fix too is that we need only want to show the browser prompt when users are [editing existing native questions](https://github.com/metabase/metabase/pull/30477). If we only rely on `question.isSave()` solely we will show the browser warning for new native questions too (`+New` > `SQL Query`) which we don't want.

I hope this is clear

WiNloSt committed:
Fix condition to show browser prompt + add tests

WiNloSt committed:
Prevent passing new props to QueryBuilder

Since those props will be pass down to other children, and they won't be
used anyway. Those new props are only used in this file.

WiNloSt committed:
Make test setup more concise

WiNloSt commented on frontend/src/metabase/query_builder/containers/QueryBuilder.jsx:
```diff
@@ -139,7 +141,6 @@ const mapStateToProps = (state, props) => {
 
     isBookmarked: getIsBookmarked(state, props),
     isDirty: getIsDirty(state),
-    isNew: getIsNew(state),
```
I've explained about the removal of `getIsNew` selector in this comment https://github.com/metabase/metabase/pull/31986/files?w=0#r1246535868. Basically, I don't think this has been used by anyone anymore.

WiNloSt committed:
Merge branch 'master' into 31974-fix-no-warning-when-leaving-edited-native-questions

ranquild commented on frontend/src/metabase/containers/SaveQuestionModal.tsx:
```diff
@@ -140,9 +141,7 @@ export default class SaveQuestionModal extends Component<SaveQuestionModalProps>
-      !question.isSaved() &&
-      !!originalQuestion &&
-      !originalQuestion.isDataset() &&
+      isQuestionEdited(question, originalQuestion) &&
```
Maybe `isSavedQuestionChanged` would be more clear

ranquild replied:
Furthermore, it shouldn't be a utility function since we can always call it as a selector. Could you make it a selector and always use as a selector?

WiNloSt replied:
Fixed in https://github.com/metabase/metabase/pull/31986/commits/1712410e50e652ef1e186be6bc7273fb6bdd67ba

WiNloSt committed:
Merge branch 'master' into 31974-fix-no-warning-when-leaving-edited-native-questions

WiNloSt commented on frontend/src/metabase/query_builder/containers/QueryBuilder.jsx:
```diff
@@ -436,3 +425,21 @@ export default _.compose(
   titleWithLoadingTime("queryStartTime"),
 )(QueryBuilder);
+
+function useShouldShowUnsavedChangesWarning() {
```
Nice catch. I didn't realize that.

WiNloSt replied:
On second thought, moving this hook to `frontend/src/metabase/query_builder/selectors.js` would require me to add unit tests for the new selector to ensure that it would work when called elsewhere. However, this feels to me that the logic only belongs to `QueryBuilder.jsx` alone, not any other component under `metabase/query_builder/*`. So leaving this here has some benefits compared to moving it to the selector file:
1. I don't need to create new tests which could be cumbersome to set up.
2. The logic is colocated. Since it's only being used in this file alone.

With the above 2 benefits, what do you think about leaving this as is?

ranquild replied:
No, I think it's incorrect to combine selectors with hooks - you should add a selector instead. Feel free to test this change in `QueryBuilder` as you do now

WiNloSt committed:
Address review: Change from util function to selector

WiNloSt commented on frontend/src/metabase/containers/SaveQuestionModal.tsx:
```diff
@@ -158,10 +160,10 @@ export const SaveQuestionModal = ({
-    !!originalQuestion &&
-    !originalQuestion.isDataset() &&
+    isSavedQuestionChanged &&
+    originalQuestion != null &&
```
Even though, `originalQuestion` would be guaranteed to have value when `isSavedQuestionChanged` is true, TypeScript won't know that, so I need to still put `originalQuestion != null` here to stop TypeScript from complaning that `originalQuestion` could be null.

WiNloSt commented:
@ranquild I've re-request your review again. There are 2 unaddressed comments since I don't think I could make changes. I've replied here.
- https://github.com/metabase/metabase/pull/31986#discussion_r1254015817
- https://github.com/metabase/metabase/pull/31986#discussion_r1254065378
Please let me know what you think.

ranquild commented on frontend/src/metabase/query_builder/selectors.js:
```diff
@@ -298,6 +297,11 @@ export const getPreviousQueryBuilderMode = createSelector(
   uiControls => uiControls.previousQueryBuilderMode,
 );
 
+export const getIsEditingModel = createSelector(
```
Ok I think that's fine. But could you not call it `isEditingModel` then? I think you could probably just use `getQueryBuilderMode` in your selector

ranquild commented on frontend/src/metabase/query_builder/selectors.unit.spec.js:
```diff
@@ -391,3 +393,49 @@ describe("getQuestionDetailsTimelineDrawerState", () => {
+const MODEL_CARD = createMockCard({ id: 1, dataset: true });
+const SAVED_QUESTION_CARD = createMockCard({ id: 2 });
+const UNSAVED_QUESTION_CARD = createMockCard({ id: undefined });
+describe("getIsSavedQuestionChanged", () => {
```
We don't test selectors as per redux docs. Test components instead

ranquild replied:
Probably your existing tests in `QueryBuilder` is fine

ranquild reviewed (commented):
Almost there. We just need to add a single selector and remove selector tests / move the checks for them in `QueryBuilder` tests

WiNloSt committed:
Merge branch 'master' into 31974-fix-no-warning-when-leaving-edited-native-questions

WiNloSt committed:
Address review: move hook to sector and move selector tests to component test

WiNloSt committed:
Fix failed unit tests from merging `master`

WiNloSt commented on frontend/src/metabase/containers/SaveQuestionModal.unit.spec.tsx:
```diff
@@ -80,6 +84,13 @@ const setup = async (
 
   const settings = mockSettings({ "enable-query-caching": isCachingEnabled });
 
+  const state = createMockState({
```
`SaveQuestionModal` now relies on a new selector, so I need to mock the state to reflect the change.

https://github.com/metabase/metabase/blob/4217c360cfe5770c1cc28e766bf6d926a4aabbe6/frontend/src/metabase/query_builder/selectors.js#L559-L569

WiNloSt commented on frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx:
```diff
@@ -286,7 +347,52 @@ describe("QueryBuilder", () => {
+    });
+  });
+
+  describe("beforeunload events in structured queries", () => {
```
This new set of tests along with https://github.com/metabase/metabase/blob/4217c360cfe5770c1cc28e766bf6d926a4aabbe6/frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx#L301

will ensure the new selector is tested thoroughly
https://github.com/metabase/metabase/blob/4217c360cfe5770c1cc28e766bf6d926a4aabbe6/frontend/src/metabase/query_builder/selectors.js#L608-L635

WiNloSt commented on frontend/src/metabase/query_builder/utils.ts:
[Originally this file is converted to TS to create a new util function which works well with TypeScript consumers](https://github.com/metabase/metabase/pull/31986/files#discussion_r1246539305), but that util is removed [following this discussion.](https://github.com/metabase/metabase/pull/31986#discussion_r1252694000)

So, eventually, the only change to this file is pure TypeScript conversion.

WiNloSt commented:
@ranquild I've addressed the rest of your comment with this commit https://github.com/metabase/metabase/pull/31986/commits/d1137829358f94a531725893fcc7b0b96ae2494e

The tests were a little tricky to setup, but they're supposed to cover the new selectors instead of relying on testing selectors directly.

WiNloSt commented:
I'm waiting for https://github.com/metabase/metabase/pull/32168 to be merged and also https://github.com/metabase/metabase/pull/31747 to be backported before this PR can be backported since it requires changes from those 2 PRs
